File: sentinel/fsm.py
"""Finite State Machine engine for tool gating."""

import json
import os
import re
import time
from dataclasses import asdict, dataclass
from pathlib import Path

from sentinel.config import FSMConfig, Transition


@dataclass
class FSMState:
    current: str
    previous: str | None = None
    entered_at: float = 0.0
    transition_count: int = 0

    def to_dict(self) -> dict:
        return asdict(self)

    @classmethod
    def from_dict(cls, d: dict) -> "FSMState":
        return cls(**d)


class SentinelFSM:
    def __init__(self, config: FSMConfig, state_path: Path) -> None:
        self._config = config
        self._state_path = state_path
        self._state = self._load_or_init()

    def _load_or_init(self) -> FSMState:
        if self._state_path.exists():
            data = json.loads(self._state_path.read_text())
            state = FSMState.from_dict(data)
            if state.current in self._config.states:
                return state
        return self._init_state()

    def _init_state(self) -> FSMState:
        state = FSMState(
            current=self._config.initial_state,
            entered_at=time.time(),
        )
        self._persist(state)
        return state

    def _persist(self, state: FSMState) -> None:
        self._state_path.parent.mkdir(parents=True, exist_ok=True)
        tmp_path = self._state_path.with_suffix(".tmp")
        tmp_path.write_text(json.dumps(state.to_dict(), indent=2))
        os.replace(str(tmp_path), str(self._state_path))

    def get_state(self) -> FSMState:
        return self._state

    def is_tool_allowed(self, tool_name: str) -> tuple[bool, str]:
        """Check if a tool is allowed in the current state.

        Returns (allowed, reason).
        """
        state_config = self._config.states[self._state.current]
        for pattern in state_config.allowed_tools:
            if re.fullmatch(pattern, tool_name):
                return True, f"Tool '{tool_name}' allowed in state '{self._state.current}'"

        return False, (
            f"Tool '{tool_name}' not allowed in state '{self._state.current}'. "
            f"Allowed: {state_config.allowed_tools}"
        )

    def evaluate_transition(self, tool_name: str, tool_input: dict) -> str | None:
        """Check if a tool call should trigger a state transition.

        Returns the target state name if a transition should fire, None otherwise.
        """
        for t in self._config.transitions:
            if not self._transition_matches_source(t):
                continue
            if t.trigger == "manual":
                continue
            if t.trigger != tool_name:
                continue
            if self._guards_pass(t, tool_input):
                return t.to_state
        return None

    def transition_to(self, state_name: str) -> FSMState:
        """Execute a state transition."""
        if state_name not in self._config.states:
            raise ValueError(f"Unknown state: '{state_name}'")

        self._state = FSMState(
            current=state_name,
            previous=self._state.current,
            entered_at=time.time(),
            transition_count=self._state.transition_count + 1,
        )
        self._persist(self._state)
        return self._state

    def get_available_transitions(self) -> list[dict]:
        """List transitions available from the current state."""
        result = []
        for t in self._config.transitions:
            if self._transition_matches_source(t):
                result.append({
                    "to": t.to_state,
                    "trigger": t.trigger,
                    "guards": [{"field": g.field, "pattern": g.pattern} for g in t.guards],
                })
        return result

    def reset(self) -> FSMState:
        """Reset to the initial state."""
        self._state = self._init_state()
        return self._state

    def _transition_matches_source(self, t: Transition) -> bool:
        return t.from_state == "*" or t.from_state == self._state.current

    def _guards_pass(self, t: Transition, tool_input: dict) -> bool:
        for guard in t.guards:
            value = tool_input.get(guard.field, "")
            if not isinstance(value, str):
                value = str(value)
            if not re.search(guard.pattern, value):
                return False
        return True

File: src/fsm.rs
use crate::config::{FsmConfig, GuardCondition, Transition};
use regex::Regex;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::{error::Error, fmt, fs, path::PathBuf, time::SystemTime};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct FsmState {
    pub current: String,
    pub previous: Option<String>,
    pub entered_at: f64,
    pub transition_count: u64,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct AvailableGuard {
    pub field: String,
    pub pattern: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct AvailableTransition {
    pub to: String,
    pub trigger: String,
    pub guards: Vec<AvailableGuard>,
}

#[derive(Debug)]
pub struct FsmError(String);

impl FsmError {
    fn new(message: impl Into<String>) -> Self {
        Self(message.into())
    }
}

impl fmt::Display for FsmError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.0)
    }
}

impl Error for FsmError {}

pub struct SentinelFsm {
    config: FsmConfig,
    state_path: PathBuf,
    state: FsmState,
}

impl SentinelFsm {
    pub fn new(config: FsmConfig, state_path: PathBuf) -> Result<Self, FsmError> {
        let state = load_or_init(&config, &state_path)?;
        Ok(Self {
            config,
            state_path,
            state,
        })
    }

    pub fn get_state(&self) -> &FsmState {
        &self.state
    }

    pub fn is_tool_allowed(&self, tool_name: &str) -> (bool, String) {
        let Some(state_config) = self.config.states.get(&self.state.current) else {
            return (
                false,
                format!("Current state '{}' is not configured", self.state.current),
            );
        };

        for pattern in &state_config.allowed_tools {
            if regex_full_match(pattern, tool_name) {
                return (
                    true,
                    format!(
                        "Tool '{tool_name}' allowed in state '{}'",
                        self.state.current
                    ),
                );
            }
        }

        (
            false,
            format!(
                "Tool '{tool_name}' not allowed in state '{}'. Allowed: {:?}",
                self.state.current, state_config.allowed_tools
            ),
        )
    }

    pub fn evaluate_transition(&self, tool_name: &str, tool_input: &Value) -> Option<String> {
        self.config
            .transitions
            .iter()
            .find(|transition| {
                self.transition_matches_source(transition)
                    && transition.trigger != "manual"
                    && transition.trigger == tool_name
                    && guards_pass(transition, tool_input)
            })
            .map(|transition| transition.to_state.clone())
    }

    pub fn transition_to(&mut self, state_name: &str) -> Result<FsmState, FsmError> {
        if !self.config.states.contains_key(state_name) {
            return Err(FsmError::new(format!("Unknown state: '{state_name}'")));
        }

        self.state = FsmState {
            current: state_name.to_string(),
            previous: Some(self.state.current.clone()),
            entered_at: now_seconds(),
            transition_count: self.state.transition_count + 1,
        };
        persist(&self.state, &self.state_path)?;
        Ok(self.state.clone())
    }

    pub fn get_available_transitions(&self) -> Vec<AvailableTransition> {
        self.config
            .transitions
            .iter()
            .filter(|transition| self.transition_matches_source(transition))
            .map(|transition| AvailableTransition {
                to: transition.to_state.clone(),
                trigger: transition.trigger.clone(),
                guards: transition
                    .guards
                    .iter()
                    .map(|guard| AvailableGuard {
                        field: guard.field.clone(),
                        pattern: guard.pattern.clone(),
                    })
                    .collect(),
            })
            .collect()
    }

    pub fn reset(&mut self) -> Result<FsmState, FsmError> {
        self.state = init_state(&self.config, &self.state_path)?;
        Ok(self.state.clone())
    }

    fn transition_matches_source(&self, transition: &Transition) -> bool {
        transition.from_state == "*" || transition.from_state == self.state.current
    }
}

fn load_or_init(config: &FsmConfig, state_path: &PathBuf) -> Result<FsmState, FsmError> {
    if state_path.exists() {
        let text = fs::read_to_string(state_path)
            .map_err(|error| FsmError::new(format!("Failed to read state: {error}")))?;
        if let Ok(state) = serde_json::from_str::<FsmState>(&text)
            && config.states.contains_key(&state.current)
        {
            return Ok(state);
        }
    }

    init_state(config, state_path)
}

fn init_state(config: &FsmConfig, state_path: &PathBuf) -> Result<FsmState, FsmError> {
    let state = FsmState {
        current: config.initial_state.clone(),
        previous: None,
        entered_at: now_seconds(),
        transition_count: 0,
    };
    persist(&state, state_path)?;
    Ok(state)
}

fn persist(state: &FsmState, state_path: &PathBuf) -> Result<(), FsmError> {
    if let Some(parent) = state_path.parent() {
        fs::create_dir_all(parent)
            .map_err(|error| FsmError::new(format!("Failed to create state dir: {error}")))?;
    }

    let tmp_path = state_path.with_extension("tmp");
    let rendered = serde_json::to_string_pretty(state)
        .map_err(|error| FsmError::new(format!("Failed to serialize state: {error}")))?;
    fs::write(&tmp_path, rendered)
        .map_err(|error| FsmError::new(format!("Failed to write state: {error}")))?;
    fs::rename(&tmp_path, state_path)
        .map_err(|error| FsmError::new(format!("Failed to replace state: {error}")))?;
    Ok(())
}

fn regex_full_match(pattern: &str, value: &str) -> bool {
    Regex::new(pattern)
        .ok()
        .and_then(|regex| regex.find(value))
        .is_some_and(|matched| matched.start() == 0 && matched.end() == value.len())
}

fn guards_pass(transition: &Transition, tool_input: &Value) -> bool {
    transition
        .guards
        .iter()
        .all(|guard| guard_matches(guard, tool_input))
}

fn guard_matches(guard: &GuardCondition, tool_input: &Value) -> bool {
    let value = tool_input
        .get(&guard.field)
        .map(value_to_string)
        .unwrap_or_default();

    Regex::new(&guard.pattern)
        .map(|regex| regex.is_match(&value))
        .unwrap_or(false)
}

fn value_to_string(value: &Value) -> String {
    match value {
        Value::String(value) => value.clone(),
        _ => value.to_string(),
    }
}

fn now_seconds() -> f64 {
    SystemTime::now()
        .duration_since(SystemTime::UNIX_EPOCH)
        .map(|duration| duration.as_secs_f64())
        .unwrap_or(0.0)
}
